editor/create_dialog.h

```diff
@@ -35,60 +35,65 @@
 #include "scene/gui/button.h"
 #include "scene/gui/dialogs.h"
 #include "scene/gui/item_list.h"
-#include "scene/gui/label.h"
 #include "scene/gui/line_edit.h"
 #include "scene/gui/tree.h"
 
 class CreateDialog : public ConfirmationDialog {
 	GDCLASS(CreateDialog, ConfirmationDialog);
 
-	Vector<String> favorite_list;
-	Tree *favorites;
-	Tree *recent;
-
-	Button *favorite;
 	LineEdit *search_box;
 	Tree *search_options;
-	HashMap<String, TreeItem *> search_options_types;
-	HashMap<String, RES> search_loaded_scripts;
-	bool is_replace_mode;
+
 	String base_type;
 	String preferred_search_result_type;
+
+	Button *favorite;
+	Vector<String> favorite_list;
+	Tree *favorites;
+	ItemList *recent;
 	EditorHelpBit *help_bit;
+
+	HashMap<String, TreeItem *> search_options_types;
+	HashMap<String, String> custom_type_parents;
+	HashMap<String, int> custom_type_indices;
 	List<StringName> type_list;
 	Set<StringName> type_blacklist;
 
 	void _item_selected();
 	void _hide_requested();
 
 	void _update_search();
-	void _update_favorite_list();
-	void _save_favorite_list();
-	void _favorite_toggled();
-
-	void _history_selected();
-	void _favorite_selected();
+	bool _should_hide_type(const String &p_type) const;
+	void _reconstruct_branch(const String &p_current, bool cpp_type);
+	void _customize_search_option_item(TreeItem *item, const String &p_type, const bool cpp_type);
+	String _top_result(const Vector<String> candidates, const String &search_text) const;
+	float _score_type(const String &p_type, const String &p_search) const;
 
-	void _history_activated();
-	void _favorite_activated();
+	void _fill_type_list();
+	void _clean_up();
 
 	void _sbox_input(const Ref<InputEvent> &p_ie);
+	void _text_changed(const String &p_newtext);
+	void select_type(const String &p_type);
 
 	void _confirmed();
-	void _text_changed(const String &p_newtext);
+	virtual void cancel_pressed();
 
-	Ref<Texture2D> _get_editor_icon(const String &p_type) const;
+	void _favorite_toggled();
 
-	void add_type(const String &p_type, HashMap<String, TreeItem *> &p_types, TreeItem *p_root, TreeItem **to_select);
+	void _history_selected(int idx);
+	void _favorite_selected();
 
-	void select_type(const String &p_type);
+	void _history_activated(int idx);
+	void _favorite_activated();
 
 	Variant get_drag_data_fw(const Point2 &p_point, Control *p_from);
 	bool can_drop_data_fw(const Point2 &p_point, const Variant &p_data, Control *p_from) const;
 	void drop_data_fw(const Point2 &p_point, const Variant &p_data, Control *p_from);
 
-	bool _is_class_disabled_by_feature_profile(const StringName &p_class);
-	bool _is_type_prefered(const String &type);
+	bool _is_class_disabled_by_feature_profile(const StringName &p_class) const;
+	bool _is_type_preferred(const String &type) const;
```

Comment:
```suggestion
	bool _is_type_preferred(const String &p_type) const;
```
(was mismatched with the `.cpp`, please check that others are consistent too)